File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/hardware/LiftInternals.java
package org.firstinspires.ftc.teamcode.hardware;

import com.qualcomm.robotcore.eventloop.opmode.OpMode;
import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.hardware.HardwareMap;
import com.qualcomm.robotcore.hardware.Servo;

import static org.firstinspires.ftc.teamcode.Utils.delay;

import org.firstinspires.ftc.teamcode.Utils;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class LiftInternals {
    public final ExecutorService liftExecutor = Executors.newSingleThreadExecutor();
    public final ExecutorService clawExecutor = Executors.newSingleThreadExecutor();
    public static final double MOTOR_SCALE_FACTOR = 0.8;
    public static final double MOTOR_UNLOCK_POWER = 0.1;
    private final OpMode opMode;
    public final DcMotor motor;
    public final Servo rotationServo;
    public final Servo clawServo;
    public final Servo lockServo;
    /** @see #setMode(DcMotor.RunMode) */
    private DcMotor.RunMode mode = DcMotor.RunMode.RUN_TO_POSITION;

    public LiftInternals(OpMode opMode) {
        this.opMode = opMode;
        HardwareMap hardwareMap = opMode.hardwareMap;
        motor = hardwareMap.get(DcMotor.class, "liftMotor");
        rotationServo = hardwareMap.get(Servo.class, "rotationServo");
        clawServo = hardwareMap.get(Servo.class, "clawServo");
        lockServo = hardwareMap.get(Servo.class, "lockServo");

        motor.setTargetPosition(Position.STACK_1.value);
        resetEncoder(); // TODO do we need this?
        motor.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);

        // Set an auto-clamp for the servo
        // These all assume that the position scaling is linear, and that we are using the center of the servo's range
        rotationServo.scaleRange(0.17, 0.845);
        clawServo.scaleRange(0.06, 0.195);
        lockServo.scaleRange(0, 0.12);
        uncheckedDrop();
    }

    // TODO for kotlin refactor grab, drop, lock, unlock, rotate into boolean vals (locked, grabbed, etc.)
    // TODO also remove the reverse functions during said refactor
    // Claw
    // Grab is 0; drop is 1
    public static final int GRAB_DELAY_MS = 700;
    public static final int DROP_DELAY_MS = 750;
    public void grab() {
        internalSetClaw(0, GRAB_DELAY_MS);
    }

    public void drop() {
        internalSetClaw(1, DROP_DELAY_MS);
    }

    // This function is NOT safe to call in a loop.
    // It is used to guarantee that the claw is ALWAYS commanded to open.
    private void uncheckedDrop() {
        setUnchecked(1, DROP_DELAY_MS);
    }

    private void internalSetClaw(int pos, long delayMillis) {
        int currentPos = (int) clawServo.getPosition();
        if (currentPos == pos) return;
        setUnchecked(pos, delayMillis);
    }

    // NOT safe for a loop
    private void setUnchecked(int pos, long delayMillis) {
        Utils.setPwm(clawServo, true);
        clawServo.setPosition(pos);
        System.out.println("position set " + clawServo.getPosition());
        clawExecutor.submit(() -> {
            Utils.delay(delayMillis);
            Utils.setPwm(clawServo, false);
        });
    }

    // Rotation
    // Grab is 1; drop is 0
    public void rotateToDrop(boolean reversed) {
        rotationServo.setPosition(reversed ? 1 : 0);
    }

    public void rotateToGrab(boolean reversed) {
        rotationServo.setPosition(reversed ? 0 : 1);
    }

    // Lock
    // Lock is 0; unlock is 1
    public void lock() {
        lockServo.setPosition(0);
    }

    public void unlock() {
        lockServo.setPosition(1);
    }

    // motor
    // Prevents constantly setting a new mode
    public void setMode(DcMotor.RunMode newMode) {
        if (mode == newMode) return;
        mode = newMode;
        motor.setMode(mode);
    }

    public void goToPositionBlocking(LiftInternals.Position targetPosition, double power) {
        // I didn't reuse as much code as I could have since I want to avoid multithreading unless needed
        boolean needsLock = goToPositionInternal(targetPosition.value, power);
        awaitSlide();
        System.out.println("motor finished moving to " + targetPosition);
        if (needsLock) lock();
    }

    /** Power MUST be positive. */
    public void goToPosition(Position targetPosition, double power) {
        goToPosition(targetPosition.value, power);
    }

    /** Power MUST be positive. */
    private void goToPosition(int targetPosition, double power) { // just in case
        liftExecutor.submit(() -> {
            if (goToPositionInternal(targetPosition, power)) {
                awaitSlide();
                System.out.println("motor finished moving to " + targetPosition);
                lock();
            }
        });
    }

    private boolean goToPositionInternal(int targetPosition, double power) {
        // positive is up
        boolean needsLock = targetPosition < motor.getCurrentPosition();
        if (needsLock) {
            setMode(DcMotor.RunMode.RUN_USING_ENCODER);
            motor.setPower(MOTOR_UNLOCK_POWER);
            delay(100);
            unlock();
        }
        delay(100);
        motor.setTargetPosition(targetPosition);
        setMode(DcMotor.RunMode.RUN_TO_POSITION);
        motor.setPower(power * MOTOR_SCALE_FACTOR);
        System.out.println("power set: " + motor.getPower());
        return needsLock;
    }

    public void awaitSlide() {
        Utils.waitUntil(() -> Math.abs(motor.getCurrentPosition() - motor.getTargetPosition()) < 50);
    }

    public void resetEncoder() {
        motor.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
        motor.setMode(mode);
    }

    public enum Position {
        // TODO document the math for why we need 5 separate stack positions
        // TODO test if we need to explicitly disable locking for the GROUND position
        // STACK_N is a stack containing N cones
        // STACK_1 is for a single cone, and should be the default bottom position
        STACK_1(0), STACK_2(170), STACK_3(300), STACK_4(400), STACK_5(510),
        // the lowest position that allows rotation
        CAN_ROTATE(1400),
        // junction heights
        LOW(1400), MIDDLE(2200), HIGH(3100);

        public static final Position GROUND = STACK_1;

        public final int value;
        Position(int value) {
            this.value = value;
        }

        public static Position getBelow(Position pos) {
            int ordinal = pos.ordinal();
            if (ordinal > 0 && ordinal < 5) { // only allow stack ones, except bottom
                return Position.values()[ordinal - 1];
            }
            return pos;
        }

        public static Position getAbove(Position pos) {
            int ordinal = pos.ordinal();
            if (ordinal < 4) { // only allow stack ones, except top
                return Position.values()[ordinal + 1];
            }
            return pos; // how did we get here?
        }
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/hardware/LiftInternals.kt
package org.firstinspires.ftc.teamcode.hardware

import org.firstinspires.ftc.teamcode.pwm
import org.firstinspires.ftc.teamcode.delay
import org.firstinspires.ftc.teamcode.waitUntil
import com.qualcomm.robotcore.eventloop.opmode.OpMode
import com.qualcomm.robotcore.hardware.DcMotor
import com.qualcomm.robotcore.hardware.Servo
import com.qualcomm.robotcore.hardware.DcMotor.RunMode
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.Future
import kotlin.math.abs
import kotlin.math.roundToInt

class LiftInternals(private val opMode: OpMode) {
    @JvmField val liftExecutor: ExecutorService = Executors.newSingleThreadExecutor()
    private val clawExecutor: ExecutorService = Executors.newSingleThreadExecutor()
    @JvmField val motor: DcMotor
    @JvmField val rotationServo: Servo
    val clawServo: Servo
    private val lockServo: Servo

    var motorMode = RunMode.RUN_TO_POSITION
        set(newMode) {
            if (field == newMode) return
            println("changing mode from $field to $newMode")
            field = newMode
            motor.mode = field
        }

    // 5 3/4, y wheel
    //

    init {
        val hardwareMap = opMode.hardwareMap
        motor = hardwareMap.dcMotor["liftMotor"]
        rotationServo = hardwareMap.servo["rotationServo"]
        clawServo = hardwareMap.servo["clawServo"]
        lockServo = hardwareMap.servo["lockServo"]
        motor.targetPosition = Position.STACK_1.value // need to have a pos set before mode is set (by resetEncoder)
        resetEncoder()
        motor.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE

        // Set an auto-clamp for the servo
        // These all assume that the position scaling is linear, and that we are using the center of the servo's range
        rotationServo.scaleRange(0.1, 0.78) // (+)
        clawServo.scaleRange(0.38, 0.8)
        lockServo.scaleRange(0.0, 0.12)

        lock() // needed to fix a bug where unlocking fails on the first cycle
        rotateToGrab()
        uncheckedDrop()
    }

    @Suppress("NOTHING_TO_INLINE")
    inline fun initSlide() = goToPosition(Position.STACK_1, 0.25)

    fun grab() {
        internalSetClaw(1, GRAB_DELAY_MS)
    }

    fun drop() {
        internalSetClaw(0, DROP_DELAY_MS)
    }

    // This function is NOT safe to call in a loop.
    // It is used to guarantee that the claw is ALWAYS commanded to open.
    internal fun uncheckedGrab() {
        setUnchecked(1, GRAB_DELAY_MS)
    }

    internal fun uncheckedDrop() {
        setUnchecked(0, DROP_DELAY_MS)
    }

    private fun internalSetClaw(pos: Int, delayMillis: Long) {
        val currentPos = clawServo.position.roundToInt()
        if (currentPos == pos) return
        setUnchecked(pos, delayMillis)
    }

    // NOT safe for a loop
    private fun setUnchecked(pos: Int, delayMillis: Long) {
        clawServo.pwm = true
        clawServo.position = pos.toDouble()
        println("position set " + clawServo.position)
        clawExecutor.submit {
            delay(delayMillis)
            clawServo.pwm = false
        }
    }

    // Rotation
    // Grab is 1; drop is 0
    fun rotateToDrop(reversed: Boolean) {
        rotationServo.position = if (reversed) 1.0 else 0.0
    }

    fun rotateToGrab() {
        rotationServo.position = 1.0
    }

    // Lock
    // Lock is 0; unlock is 1
    fun lock() {
        lockServo.position = 0.0
    }

    fun unlock() {
        lockServo.position = 1.0
    }

    // motor
    @Suppress("NOTHING_TO_INLINE")
    inline fun goToPositionBlocking(targetPosition: Position, power: Double) {
        goToPositionBlocking(targetPosition.value, power)
    }

    fun goToPositionBlocking(pos: Int, power: Double) {
        // I didn't reuse as much code as I could have since I want to avoid multithreading unless needed
        val needsLock = goToPositionInternal(pos, power)
        awaitTargetHit()
        println("motor finished moving to $pos")
        if (needsLock) lock()
    }

    /** Power MUST be positive.  */
    fun goToPosition(targetPosition: Position, power: Double): Future<*> {
        return goToPosition(targetPosition.value, power)
    }

    /** Power MUST be positive.  */
    fun goToPosition(targetPosition: Int, power: Double): Future<*> { // just in case
        return liftExecutor.submit {
            val needsLock = goToPositionInternal(targetPosition, power)
            awaitTargetHit()
            println("motor finished moving to $targetPosition")
            if (needsLock) lock()
        }
    }

    private fun goToPositionInternal(targetPosition: Int, power: Double): Boolean {
        // positive is up
        val needsLock = targetPosition < motor.currentPosition
        if (needsLock) {
            motorMode = RunMode.RUN_USING_ENCODER
            motor.power = MOTOR_UNLOCK_POWER
            delay(LOCK_DELAY_MS)
            unlock()
        }

        delay(100)

        motor.targetPosition = targetPosition
        motorMode = RunMode.RUN_TO_POSITION
        motor.power = power * MOTOR_SCALE_FACTOR

        println("power set: " + motor.power)
        return needsLock
    }

    @Suppress("NOTHING_TO_INLINE")
    inline fun awaitTargetHit() {
        waitUntil { abs(motor.currentPosition - motor.targetPosition) < 50 }
    }

    fun awaitClaw() {
        waitUntil { !clawServo.pwm } // only true while powered
    }

    fun resetEncoder() {
        motor.mode = RunMode.STOP_AND_RESET_ENCODER
        motor.mode = motorMode
    }

    enum class Position(@JvmField val value: Int) {
        // TODO document the math for why we need 5 separate stack positions
        // TODO test if we need to explicitly disable locking for the GROUND position
        // STACK_N is a stack containing N cones
        // STACK_1 is for a single cone, and should be the default bottom position

        STACK_1(210), STACK_2(260), STACK_3(350), STACK_4(490), STACK_5(630),  // the lowest position that allows rotation

        CAN_ROTATE(1530),  // junction heights

        LOW(1530), MIDDLE(2250), HIGH(3100),

        ZERO(0);

        operator fun inc(): Position {
            val ordinal = ordinal
            return if (ordinal < 4) { // only allow stack ones, except top
                values()[ordinal + 1]
            } else this // how did we get here?
        }

        operator fun dec(): Position {
            val ordinal = ordinal
            return if (ordinal in 1..4) { // only allow stack ones, except bottom
                values()[ordinal - 1]
            } else this
        }
    }

    companion object {
        const val MOTOR_SCALE_FACTOR = 0.8
        const val MOTOR_UNLOCK_POWER = 0.2
        const val LOCK_DELAY_MS = 200L

        // TODO for kotlin refactor grab, drop, lock, unlock, rotate into boolean vals (locked, grabbed, etc.)
        // TODO also remove the reverse functions during said refactor
        // Claw
        // Grab is 0; drop is 1
        const val GRAB_DELAY_MS = 600L
        const val DROP_DELAY_MS = 700L
    }
}